bpo-37879: Suppress subtype_dealloc decref when base type is a C heap type

eduardo-elizondo opened the pull request:
The instance destructor for a type is responsible for preparing
an instance for deallocation by decrementing the reference counts
of its referents.

If an instance belongs to a heap type, the type object of an instance
has its reference count decremented while for static types, which
are permanently allocated, the type object is unaffected by the
instance destructor.

Previously, the default instance destructor searched the class
hierarchy for an inherited instance destructor and, if present,
would invoke it.

Then, if the instance type is a heap type, it would decrement the
reference count of that heap type.  However, this could result in the
premature destruction of a type because the inherited instance
destructor should have already decremented the reference count
of the type object.

This change avoids the premature destruction of the type object
by suppressing the decrement of its reference count when an
inherited, non-default instance destructor has been invoked.

Finally, an assertion on the Py_SIZE of a type was deleted.  Heap
types have a non zero size, making this into an incorrect assertion.

<!-- issue-number: [bpo-37879](https://bugs.python.org/issue37879) -->
https://bugs.python.org/issue37879
<!-- /issue-number -->

eduardo-elizondo committed:
Suppress subtype_dealloc decref when base type is a C heap type

The instance destructor for a type is responsible for preparing
an instance for deallocation by decrementing the reference counts
of its referents.

If an instance belongs to a heap type, the type object of an instance
has its reference count decremented while for static types, which
are permanently allocated, the type object is unaffected by the
instance destructor.

Previously, the default instance destructor searched the class
hierarchy for an inherited instance destructor and, if present,
would invoke it.

Then, if the instance type is a heap type, it would decrement the
reference count of that heap type.  However, this could result in the
premature destruction of a type because the inherited instance
destructor should have already decremented the reference count
of the type object.

This change avoids the premature destruction of the type object
by suppressing the decrement of its reference count when an
inherited, non-default instance destructor has been invoked.

Finally, an assertion on the Py_SIZE of a type was deleted.  Heap
types have a non zero size, making this into an incorrect assertion.

eduardo-elizondo committed:
added blurb

eduardo-elizondo committed:
Fixed whitespace

kevinjacobs-progenity commented on Objects/typeobject.c:
```diff
@@ -1157,11 +1157,9 @@ subtype_dealloc(PyObject *self)
-           adding any slots or instance variables.  This allows
-           certain simplifications: there's no need to call
-           clear_slots(), or DECREF the dict, or clear weakrefs. */
+        /* A non GC dyanmic type allows certain simplifications:
```
typo in dynamic

eduardo-elizondo replied:
Thanks, fixed!

kevinjacobs-progenity commented on Objects/typeobject.c:
```diff
@@ -1189,8 +1186,10 @@ subtype_dealloc(PyObject *self)
+       /* Only decref if the base type is not already a heap allocated type.
+          Otherwise, basedealloc should have decref'd it already */
+        if (type->tp_flags & Py_TPFLAGS_HEAPTYPE && !(base->tp_flags & Py_TPFLAGS_HEAPTYPE))
+                                                Py_DECREF(type);
```
odd looking indent, at least via github

eduardo-elizondo replied:
You are right that was definitely non-intentional. I think my auto rebase messed this up. Anyways, fixed!

eduardo-elizondo committed:
Typo and indent

jdemeyer commented on Objects/typeobject.c:
```diff
@@ -1189,8 +1186,10 @@ subtype_dealloc(PyObject *self)
-        Py_DECREF(type);
+       /* Only decref if the base type is not already a heap allocated type.
+          Otherwise, basedealloc should have decref'd it already */
+        if (type->tp_flags & Py_TPFLAGS_HEAPTYPE && !(base->tp_flags & Py_TPFLAGS_HEAPTYPE))
```
Why the first condition `type->tp_flags & Py_TPFLAGS_HEAPTYPE`? There is already an assertion at the beginning of `subtype_dealloc`.

eduardo-elizondo replied:
The type can be changed by `tp_del`. If you look closely at the implementation, the type is extracted again. So the type that you are looking at this spot might not be the same type that you started off with.

jdemeyer replied:
> The type can be changed by `tp_del`

I see. Reading the code for `__class__` assignment, this would only be possible for subtypes of `module`. Do you know if there is a testcase for this?

eduardo-elizondo replied:
It's possible for both subtypes of module and heap types. The specific check done by `object_set_class` in `typeobject.c` is: https://github.com/python/cpython/blob/master/Objects/typeobject.c#L4049-L4057

And yes, I've added two tests for this. One for the `!PyType_IS_GC` case (which is the one you have here). And for the `PyType_IS_GC` case down below.

Both tests have been added to `test_capi.py` and are named `test_c_subclass_of_heap_ctype_with_del_modifying_dunder_class_only_decrefs_once` and `test_subclass_of_heap_gc_ctype_with_del_modifying_dunder_class_only_decrefs_once` respectively.

jdemeyer reviewed (changes requested):
Apart from the one comment, the changes to `typeobject.c` look good. I haven't yet looked at the testcases.

encukou commented on Modules/_testcapimodule.c:
```diff
@@ -6012,6 +6012,180 @@ static struct PyModuleDef _testcapimodule = {
+    PyTypeObject *base = (PyTypeObject *)PyType_GetSlot(ctypesubclass, Py_tp_base);
+    Py_DECREF(ctypesubclass);
+    initproc base_init = PyType_GetSlot(base, Py_tp_init);
+    if (base_init(self, args, kwargs) < 0) {
```
Is this setup to get the superclass necessary?
It should be possible to call `heapctype_init(self, args, kwargs)` directly.

eduardo-elizondo replied:
It is indeed. I actually had your suggestion initially but ultimately ended up changing it to this approach. My rationale was: To show that `HeapCTypeSubclass` is indeed a C extension subclass and that you can call the `tp_init` slot by pulling its `tp_base` out. To me, it seemed closer to what a `__init__` would do by calling `super().__init__`.

That being said, I don't have any strong opinions on either. I'll keep this as is but feel free to add one more comment to change to to call `heapctype_init` and I will change it this time!

encukou replied:
There are some subtle issues in doing dynamic `super()`-like calls from C code, but using `PyState_FindModule`/`PyObject_GetAttrString` is especially fragile.
If the elaborate code manages to set `base_init` to anything except `heapctype_init`, it's a bug. It's not the bug this is supposed to be testing; and it's a bug that's likely to segfault.

I'm actually trying to provide a better way to do this without `PyState_FindModule` ([PEP 573](https://www.python.org/dev/peps/pep-0573/)), but so far, using `heapctype_init` is, IMO, best.

encukou commented on Modules/_testcapimodule.c:
```diff
@@ -6012,6 +6012,180 @@ static struct PyModuleDef _testcapimodule = {
+
+    m = PyState_FindModule(&_testcapimodule);
+    if (m == NULL) {
+        return;
```
This leaks error_type, error_value, error_traceback.

eduardo-elizondo replied:
You are right. Let me fix this.

Edit: Fixed

encukou committed:
Avoid PyState_FindModule in heapctypesubclass_init

encukou committed:
Move the added test code

- Move CAPITest additions to the end (they're testing a fairly
  exotic edge case, so I wouldn't make them prominent).
- Have the PyModuleDef for _testcapi close to the PyInit function again
  (this needs a forward declaration, but keeps the module init code
  logically together).

encukou committed:
Add docstrings to C classes

This is mainly to make the intention clearer to readers of the code:
the docstrings serve as headings.

encukou committed:
Remove unnecessary descriptor

The same descriptor is in the superclass.

eduardo-elizondo committed:
Fix leaks

eduardo-elizondo reviewed (commented):
LGTM

eduardo-elizondo reviewed (commented):
Both changes and the reasoning looks good to me

eduardo-elizondo reviewed (commented):
Nice catch!

eduardo-elizondo reviewed (commented):
Looks good

eduardo-elizondo committed:
Merge branch 'master' into fix-cheaptype-subclass

eduardo-elizondo committed:
Remove merge artifacts
